libraries/AC_Avoidance/AP_Astar_RT.cpp

```diff
@@ -0,0 +1,686 @@
+
+        // skip if empty
+        const AP_OAVisGraph &curr_visgraph = *visgraphs[v];
+        if (curr_visgraph.num_items() == 0) {
+            continue;
+        }
+
+        // search visibility graph for items visible from current_node
+        for (uint16_t i = 0; i < curr_visgraph.num_items(); i++) {
+            const AP_OAVisGraph::VisGraphItem &item = curr_visgraph[i];
+            // match if current node's id matches either of the id's in the graph (i.e. either end of the vector)
+            if ((curr_node.id == item.id1) || (curr_node.id == item.id2)) {
+                AP_OAVisGraph::OAItemID matching_id = (curr_node.id == item.id1) ? item.id2 : item.id1;
+                // find item's id in node array
+                node_index item_node_idx;
+                if (find_node_from_id(matching_id, item_node_idx)) {
+                    // if current node's distance + distance to item is less than item's current distance, update item's distance
+                    const float dist_to_item_via_current_node = _short_path_data[curr_node_idx].distance_cm + item.distance_cm;
+                    if (dist_to_item_via_current_node < _short_path_data[item_node_idx].distance_cm) {
+                        // update item's distance and set "distance_from_idx" to current node's index
+                        _short_path_data[item_node_idx].distance_cm = dist_to_item_via_current_node;
+                        _short_path_data[item_node_idx].distance_from_idx = curr_node_idx;
+                    }
+                }
+            }
+        }
+    }
+}
+
+#if OA_RTASTAR_DEBUG_ENABLE
+// send debug info as ADSB messages
+void AP_OART_AStar::send_debug_info(mavlink_channel_t chan, uint16_t interval_ms)
+{
+    const char callsign[9] = "OA_NODE";
+    for (uint16_t i=0; i<_total_oa_db_numpoints; i++) {
+        const Location item_loc(oa_db_nodes[i], Location::AltFrame::ABOVE_ORIGIN);
+        mavlink_msg_adsb_vehicle_send(chan,
+            i,
+            item_loc.lat,
+            item_loc.lng,
+            0,                          // altitude_type
+            item_loc.alt,
+            0,                          // heading
+            0,                          // hor_velocity
+            0,                          // ver_velocity
+            callsign,                   // callsign
+            255,                        // emitter_type
+            0,                          // tslc
+            0,                          // flags
+            0   // squawk
+        );
+    }
+}
+#endif
+
+// return point from final path as an offset (in cm) from the ekf origin
+bool AP_OART_AStar::get_shortest_path_point(uint16_t point_num, Vector3f& pos)
+{
+    if ((_path_numpoints == 0) || (point_num >= _path_numpoints)) {
+        return false;
+    }
+
+    // get id from path
+    AP_OAVisGraph::OAItemID id = _path[_path_numpoints - point_num - 1];
+
+    return convert_node_to_point(id, pos);
+}
+
+// singleton instance
```

Comment:
I wonder why the new AStar has a singleton but Dijkstra's doesn't.  Do we need it?  There's nothing really wrong with it but let's not add it if we don't need it.